### ISSUE TITLE
Artisan Exiting with Code 255; Composer Artisan Discovery Memory Leak

### ISSUE THREAD

**Issue by dfoster-cb911** [aschmelyun/docker-compose-laravel#64]
I've been running this awesome setup you have here for about 4-5 months now and it has always worked without a hitch, but today randomly, with no changes to any of the setup you have here, artisan seems to just be not working at all, period. 

It was working 100% fine yesterday when I was working the app, and I spun it down using the typical `docker-compose down --remove-orphans --v` like I've always done with this. 

Today coming back and running `docker-compose up -d --build` like I always do to spin it up works just fine, but for some reason now when I run: `docker-compose run --rm artisan migrate` it just quits out with `255` exit code and no error message at all. 

Here is what happens after running `docker-compose --verbose run --rm artisan`, for example, which had always shown me the artisan available commands before today: 

```
...
Starting mysql ...
compose.parallel.feed_queue: Pending: {<Container: mysql (b7485f)>}
compose.parallel.feed_queue: Starting producer thread for <Container: mysql (b7485f)>
Starting mysql ... done
compose.parallel.feed_queue: Pending: set()
compose.parallel.parallel_execute_iter: Finished processing: <Service: mysql>
compose.parallel.feed_queue: Pending: set()
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- ('odysseus-api_laravel')
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/networks/odysseus-api_laravel HTTP/1.1" 200 1290
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {'Attachable': True,
 'ConfigFrom': {'Network': ''},
 'ConfigOnly': False,
 'Containers': {'8fb9376af30d40eb75de066e6b71dfe411133f1d80cdb363ffb522fa5f65eb03': {'EndpointID': 'f3864f8c85956ce421af5399fb6f76dcad08b743a9f9fd38f17aef386217d85f',
                                                                                     'IPv4Address': '172.18.0.2/16',
                                                                                     'IPv6Address': '',
                                                                                     'MacAddress': '02:42:ac:12:00:02',
                                                                                     'Name': 'php'},
                'a0ed964f879ea0a1fa7193ee6e218b9e287fa983ceead5d1b2f7e20246b8ebc3': {'EndpointID': '7576f1a58432fc8a89abdf9b2caef47fe4c4068b1af7bc3db36a0208f793bf2d',
                                                                                     'IPv4Address': '172.18.0.5/16',
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- ('odysseus-api_default')
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/networks/odysseus-api_default HTTP/1.1" 200 561
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {'Attachable': True,
 'ConfigFrom': {'Network': ''},
 'ConfigOnly': False,
 'Containers': {},
 'Created': '2020-08-30T18:33:29.7785226Z',
 'Driver': 'bridge',
 'EnableIPv6': False,
 'IPAM': {'Config': [{'Gateway': '172.19.0.1', 'Subnet': '172.19.0.0/16'}],
          'Driver': 'default',
          'Options': None},
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- ('odysseus-api_artisan')
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/images/odysseus-api_artisan/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {'Architecture': 'amd64',
 'Author': '',
 'Comment': '',
 'Config': {'ArgsEscaped': True,
            'AttachStderr': False,
            'AttachStdin': False,
            'AttachStdout': False,
            'Cmd': ['php-fpm'],
            'Domainname': '',
            'Entrypoint': ['docker-php-entrypoint'],
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=False, filters={'label': ['com.docker.compose.project=odysseus-api', 'com.docker.compose.service=artisan', 'com.docker.compose.oneoff=False']})
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/json?limit=-1&all=0&size=0&trunc_cmd=0&filters=%7B%22label%22%3A+%5B%22com.docker.compose.project%3Dodysseus-api%22%2C+%22com.docker.compose.service%3Dartisan%22%2C+%22com.docker.compose.oneoff%3DFalse%22%5D%7D HTTP/1.1" 200 3
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=False, filters={'label': ['com.docker.compose.project=odysseusapi', 'com.docker.compose.service=artisan', 'com.docker.compose.oneoff=False']})
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/json?limit=-1&all=0&size=0&trunc_cmd=0&filters=%7B%22label%22%3A+%5B%22com.docker.compose.project%3Dodysseusapi%22%2C+%22com.docker.compose.service%3Dartisan%22%2C+%22com.docker.compose.oneoff%3DFalse%22%5D%7D HTTP/1.1" 200 3
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker create_host_config <- (links=[], port_bindings={}, binds=['/home/dfoster/Code/odysseus-api/src:/var/www/html:rw'], volumes_from=[], privileged=False, network_mode='odysseus-api_laravel', devices=None, dns=None, dns_opt=None, dns_search=None, restart_policy=None, runtime=None, cap_add=None, cap_drop=None, mem_limit=None, mem_reservation=None, memswap_limit=None, ulimits=None, log_config={'Type': '', 'Config': {}}, extra_hosts=None, read_only=None, pid_mode=None, security_opt=None, ipc_mode=None, cgroup_parent=None, cpu_quota=None, shm_size=None, sysctls=None, pids_limit=None, tmpfs=[], oom_kill_disable=None, oom_score_adj=None, mem_swappiness=None, group_add=None, userns_mode=None, init=None, init_path=None, isolation=None, cpu_count=None, cpu_percent=None, nano_cpus=None, volume_driver=None, cpuset_cpus=None, cpu_shares=None, storage_opt=None, blkio_weight=None, blkio_weight_device=None, device_read_bps=None, device_read_iops=None, device_write_bps=None, device_write_iops=None, mounts=None, device_cgroup_rules=None, cpu_period=None, cpu_rt_period=None, cpu_rt_runtime=None)
compose.cli.verbose_proxy.proxy_callable: docker create_host_config -> {'Binds': ['/home/dfoster/Code/odysseus-api/src:/var/www/html:rw'],
 'Links': [],
 'LogConfig': {'Config': {}, 'Type': ''},
 'NetworkMode': 'odysseus-api_laravel',
 'PortBindings': {},
 'VolumesFrom': []}
compose.cli.verbose_proxy.proxy_callable: docker create_container <- (entrypoint=['php', '/var/www/html/artisan'], volumes={'/var/www/html': {}}, working_dir='/var/www/html', command=None, tty=True, stdin_open=True, detach=False, ports=[], name='odysseus-api_artisan_run_2eaed7d9b219', environment=[], labels={'com.docker.compose.project': 'odysseus-api', 'com.docker.compose.service': 'artisan', 'com.docker.compose.oneoff': 'True', 'com.docker.compose.project.working_dir': '/home/dfoster/Code/odysseus-api', 'com.docker.compose.project.config_files': 'docker-compose.yml', 'com.docker.compose.slug': '2eaed7d9b2190bb0076e6ad931c9880eebf1904f90a0ceae39cfe0d8b2e1ae0', 'com.docker.compose.version': '1.26.2'}, image='odysseus-api_artisan', host_config={'NetworkMode': 'odysseus-api_laravel', 'VolumesFrom': [], 'Binds': ['/home/dfoster/Code/odysseus-api/src:/var/www/html:rw'], 'PortBindings': {}, 'Links': [], 'LogConfig': {'Type': '', 'Config': {}}}, networking_config={'EndpointsConfig': {'odysseus-api_laravel': {'Aliases': ['artisan'], 'IPAMConfig': {}}}})
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/containers/create?name=odysseus-api_artisan_run_2eaed7d9b219 HTTP/1.1" 201 88
compose.cli.verbose_proxy.proxy_callable: docker create_container -> {'Id': '076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905',
 'Warnings': []}
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- ('076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905')
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {'AppArmorProfile': '',
 'Args': ['/var/www/html/artisan'],
 'Config': {'AttachStderr': True,
            'AttachStdin': True,
            'AttachStdout': True,
            'Cmd': None,
            'Domainname': '',
            'Entrypoint': ['php', '/var/www/html/artisan'],
            'Env': ['PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                    'PHPIZE_DEPS=autoconf \t\tdpkg-dev dpkg \t\tfile \t\t'
...
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network <- ('076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905', 'odysseus-api_laravel')
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/networks/odysseus-api_laravel/disconnect HTTP/1.1" 200 0
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network -> None
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network <- ('076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905', 'odysseus-api_laravel', aliases=[], ipv4_address=None, ipv6_address=None, links=[], link_local_ips=None)
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/networks/odysseus-api_laravel/connect HTTP/1.1" 200 0
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network -> None
compose.cli.main.call_docker: /usr/bin/docker start --attach --interactive 076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905
compose.cli.verbose_proxy.proxy_callable: docker remove_container <- ('076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905', force=True, v=True)
urllib3.connectionpool._make_request: http://localhost:None "DELETE /v1.25/containers/076569cb74c898181d5b4d3f790d69c767a05d6fff169a8dee6abdadb7847905?v=True&link=False&force=True HTTP/1.1" 204 0
compose.cli.verbose_proxy.proxy_callable: docker remove_container -> None
```

I'm currently running this through WSL 2 in Windows: 

```
PS C:\Some\Path\To\User\dfost> wsl --list --v
  NAME                   STATE           VERSION
* Ubuntu-20.04           Running         2
  docker-desktop         Running         2
  docker-desktop-data    Running         2
```

**Composer** container seems to spin up and run its command JUST FINE however during the running of `@php artisan package:discover --ansi` via composer auto scripts ( package discovery) , the artisan line there just causes a MASSIVE memory leak which ends up eating up all my RAM in a matter of minutes. 

I'm not sure if any of your stuff here causes this, but it's frustrating as this seems to have just kind of randomly started occurring with no changes to the config. Wondering if there is any chance for guidance/ideas? 

Lmk if more information is needed.

## UPDATE:

Also, this is what I get for just trying to spin that artisan container up in general: 

```
docker-compose --verbose up artisan
```
**Output:**

```
Starting artisan ...
compose.parallel.feed_queue: Pending: {<Container: artisan (2005f8)>}
compose.parallel.feed_queue: Starting producer thread for <Container: artisan (2005f8)>
compose.cli.verbose_proxy.proxy_callable: docker attach <- ('2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017', stdout=True, stderr=True, stream=True)
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/containers/2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017/attach?logs=0&stdout=1&stderr=1&stream=1 HTTP/1.1" 101 0
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker attach -> <docker.types.daemon.CancellableStream object at 0x7f0d2c5868d0>
compose.cli.verbose_proxy.proxy_callable: docker start <- ('2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017')
compose.parallel.feed_queue: Pending: set()
compose.parallel.feed_queue: Pending: set()
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/containers/2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017/start HTTP/1.1" 204 0
compose.cli.verbose_proxy.proxy_callable: docker start -> None
Starting artisan ... done
compose.parallel.feed_queue: Pending: set()
compose.parallel.parallel_execute_iter: Finished processing: <Service: artisan>
compose.parallel.feed_queue: Pending: set()
Attaching to artisan
compose.cli.verbose_proxy.proxy_callable: docker events <- (filters={'label': ['com.docker.compose.project=odysseus-api', 'com.docker.compose.oneoff=False']}, decode=True)
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/events?filters=%7B%22label%22%3A+%5B%22com.docker.compose.project%3Dodysseus-api%22%2C+%22com.docker.compose.oneoff%3DFalse%22%5D%7D HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker events -> <docker.types.daemon.CancellableStream object at 0x7f0d2c51a2d0>
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- ('2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017')
compose.cli.verbose_proxy.proxy_callable: docker wait <- ('2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017')
urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/containers/2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017/json HTTP/1.1" 200 None
urllib3.connectionpool._make_request: http://localhost:None "POST /v1.25/containers/2005f8039c97b7e855c22dc9df2f333af94c110c2e1548dfafd0a4d8cf2ef017/wait HTTP/1.1" 200 32
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {'AppArmorProfile': '',
 'Args': ['/var/www/html/artisan'],
 'Config': {'AttachStderr': False,
            'AttachStdin': False,
            'AttachStdout': False,
            'Cmd': None,
            'Domainname': '',
            'Entrypoint': ['php', '/var/www/html/artisan'],
            'Env': ['PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                    'PHPIZE_DEPS=autoconf \t\tdpkg-dev dpkg \t\tfile \t\t'
...
compose.cli.verbose_proxy.proxy_callable: docker wait -> {'Error': None, 'StatusCode': 255}
artisan exited with code 255
```

**Comment by luc0** [comment]
I've the same issue, artisan is not working anymore.
```
artisan exited with code 255
```
@dfoster-cb911 Did you resolve it?

**Comment by dfoster-cb911** [comment]
Hey @luc0 , you'll have to excuse me for not coming back here with my findings after I had fixed it. 

My issue actually boiled down to a circular reference I had when doing some dependency injection for some new classes I had created through a new Provider class.

I wasn't quite able to figure out why there was no verbosity with the root cause in the output of the artisan error logs, but I'm glad you found out what it was on your end. 

Error `255` seems to be what is thrown when there is no idea how to handle the processing of that error, so it hides the true error and kills with a general 'There was an error' error code. I may take some personal time to dig into how I might handle the error handling, even if only to make sure the root error was put into std out logs.
